File: src/models/Position.ts
import { PositionData } from '../types/stock';

/**
 * Object-Oriented Position Model
 * Represents shares owned of a single stock, cost basis, and calculated profits.
 */
export class Position {
  private _symbol: string;
  private _shares: number;
  private _totalInvested: number;

  constructor(symbol: string, shares: number = 0, totalInvested: number = 0) {
    this._symbol = symbol;
    this._shares = shares;
    this._totalInvested = totalInvested;
  }

  public get symbol(): string { return this._symbol; }
  public get shares(): number { return this._shares; }
  public get totalInvested(): number { return this._totalInvested; }

  public get avgBuyPrice(): number {
    if (this._shares <= 0) return 0;
    return Number((this._totalInvested / this._shares).toFixed(2));
  }

  /**
   * Adds bought shares to the position, updating the cost basis
   */
  public addShares(count: number, executionPrice: number): void {
    if (count <= 0) throw new Error('Shares to add must be positive');
    this._shares += count;
    this._totalInvested += count * executionPrice;
  }

  /**
   * Reduces shares from the position and returns the cost basis portion removed
   */
  public removeShares(count: number): number {
    if (count <= 0) throw new Error('Shares to remove must be positive');
    if (count > this._shares) throw new Error('Cannot sell more shares than currently owned');

    const costBasisPortion = this.avgBuyPrice * count;
    this._shares -= count;
    this._totalInvested = Math.max(0, this._totalInvested - costBasisPortion);
    return costBasisPortion;
  }

  public calculateMetrics(currentMarketPrice: number): PositionData {
    const currentValue = Number((this._shares * currentMarketPrice).toFixed(2));
    const unrealizedPnL = Number((currentValue - this._totalInvested).toFixed(2));
    const unrealizedPnLPercent = this._totalInvested > 0
      ? Number(((unrealizedPnL / this._totalInvested) * 100).toFixed(2))
      : 0;

    return {
      symbol: this._symbol,
      shares: this._shares,
      avgBuyPrice: this.avgBuyPrice,
      totalInvested: Number(this._totalInvested.toFixed(2)),
      currentPrice: currentMarketPrice,
      currentValue,
      unrealizedPnL,
      unrealizedPnLPercent,
    };
  }

  public toJSON(): { symbol: string; shares: number; totalInvested: number } {
    return {
      symbol: this._symbol,
      shares: this._shares,
      totalInvested: Number(this._totalInvested.toFixed(2)),
    };
  }
}

File: java/com/trading/model/Position.java
package com.trading.model;

import java.io.Serializable;

public class Position implements Serializable {
    private static final long serialVersionUID = 1L;

    private final String symbol;
    private int shares;
    private double totalInvested;

    public Position(String symbol, int shares, double totalInvested) {
        if (shares < 0 || totalInvested < 0) {
            throw new IllegalArgumentException("Shares and total invested must be non-negative");
        }
        this.symbol = symbol.toUpperCase();
        this.shares = shares;
        this.totalInvested = totalInvested;
    }

    public String getSymbol() { return symbol; }
    public int getShares() { return shares; }
    public double getTotalInvested() { return totalInvested; }

    public double getAverageBuyPrice() {
        if (shares == 0) return 0.0;
        return Math.round((totalInvested / shares) * 100.0) / 100.0;
    }

    public void addShares(int count, double pricePerShare) {
        if (count <= 0) throw new IllegalArgumentException("Share count must be positive");
        this.shares += count;
        this.totalInvested += (count * pricePerShare);
    }

    public double removeShares(int count) {
        if (count <= 0) throw new IllegalArgumentException("Share count must be positive");
        if (count > this.shares) throw new IllegalArgumentException("Cannot sell more shares than currently owned");

        double costBasisRemoved = getAverageBuyPrice() * count;
        this.shares -= count;
        this.totalInvested = Math.max(0.0, this.totalInvested - costBasisRemoved);
        return costBasisRemoved;
    }

    public double calculateMarketValue(double currentPrice) {
        return Math.round(shares * currentPrice * 100.0) / 100.0;
    }

    public double calculateUnrealizedPnL(double currentPrice) {
        return Math.round((calculateMarketValue(currentPrice) - totalInvested) * 100.0) / 100.0;
    }

    public double calculateUnrealizedPnLPercent(double currentPrice) {
        if (totalInvested == 0) return 0.0;
        return Math.round((calculateUnrealizedPnL(currentPrice) / totalInvested) * 10000.0) / 100.0;
    }

    @Override
    public String toString() {
        return String.format("%s: %d shares @ Avg $%.2f | Invested: $%.2f",
                symbol, shares, getAverageBuyPrice(), totalInvested);
    }
}
